File: intelligence/cloudflare/worker/catalog.ts
import { ApiError, enumString, jsonText, optionalBoolean, optionalString, requireArray, requireNumber, requireObject, requireString } from "./http.ts";
import type { ApiResponse, AuthContext, D1PreparedStatement, JsonObject } from "./types.ts";

const TARGET_TYPES = ["company", "product", "person", "project", "topic"] as const;
const PRIORITIES = ["low", "normal", "high", "critical"] as const;

function boolInt(value: unknown, fallback: boolean): number {
  return optionalBoolean(value, fallback) ? 1 : 0;
}

export async function getCatalog({ env }: AuthContext): Promise<ApiResponse> {
  const statements = [
    env.DB.prepare("SELECT * FROM targets ORDER BY slug"),
    env.DB.prepare("SELECT * FROM channels ORDER BY target_id, slug"),
    env.DB.prepare("SELECT * FROM tags ORDER BY tag_type, slug"),
    env.DB.prepare("SELECT target_id, tag_id FROM target_tags ORDER BY target_id, tag_id"),
    env.DB.prepare("SELECT channel_id, tag_id FROM channel_tags ORDER BY channel_id, tag_id"),
  ];
  const [targets, channels, tags, targetTags, channelTags] = await env.DB.batch(statements);
  return {
    status: 200,
    body: {
      targets: targets.results ?? [],
      channels: channels.results ?? [],
      tags: tags.results ?? [],
      target_tags: targetTags.results ?? [],
      channel_tags: channelTags.results ?? [],
    },
  };
}

export async function syncCatalog({ env, body }: AuthContext): Promise<ApiResponse> {
  const envelope = requireObject(body);
  const payload = envelope.catalog === undefined ? envelope : requireObject(envelope.catalog, "catalog");
  const now = new Date().toISOString();
  const mode = payload.mode === undefined ? "replace" : enumString(payload.mode, "mode", ["replace", "merge"]);
  const targets = requireArray(payload.targets, "targets", 500).map((entry, index) => {
    const value = requireObject(entry, `targets[${index}]`);
    return {
      id: requireString(value.id, `targets[${index}].id`, { max: 128 })!,
      slug: requireString(value.slug, `targets[${index}].slug`, { max: 128 })!,
      name: requireString(value.name, `targets[${index}].name`, { max: 256 })!,
      targetType: enumString(value.target_type ?? value.type, `targets[${index}].target_type`, TARGET_TYPES),
      description: optionalString(value.description, `targets[${index}].description`, 2_000),
      priority: value.priority === undefined ? "normal" : enumString(value.priority, `targets[${index}].priority`, PRIORITIES),
      enabled: boolInt(value.enabled, true),
      createdAt: optionalString(value.created_at, `targets[${index}].created_at`, 64) ?? now,
    };
  });
  const tags = requireArray(payload.tags, "tags", 1_000).map((entry, index) => {
    const value = requireObject(entry, `tags[${index}]`);
    return {
      id: requireString(value.id, `tags[${index}].id`, { max: 128 })!,
      slug: requireString(value.slug, `tags[${index}].slug`, { max: 128 })!,
      name: requireString(value.name, `tags[${index}].name`, { max: 256 })!,
      tagType: requireString(value.tag_type ?? value.type, `tags[${index}].tag_type`, { max: 64 })!,
      createdAt: optionalString(value.created_at, `tags[${index}].created_at`, 64) ?? now,
    };
  });
  const channels = requireArray(payload.channels, "channels", 2_000).map((entry, index) => {
    const value = requireObject(entry, `channels[${index}]`);
    return {
      id: requireString(value.id, `channels[${index}].id`, { max: 128 })!,
      targetId: requireString(value.target_id, `channels[${index}].target_id`, { max: 128 })!,
      slug: requireString(value.slug, `channels[${index}].slug`, { max: 128 })!,
      name: requireString(value.name, `channels[${index}].name`, { max: 256 })!,
      channelType: requireString(value.channel_type ?? value.type, `channels[${index}].channel_type`, { max: 64 })!,
      collectorType: requireString(value.collector_type ?? value.collector, `channels[${index}].collector_type`, { max: 64 })!,
      url: optionalString(value.url, `channels[${index}].url`, 4_096),
      handle: optionalString(value.handle, `channels[${index}].handle`, 256),
      intervalMinutes: value.interval_minutes === undefined ? 60 : requireNumber(value.interval_minutes, `channels[${index}].interval_minutes`, 5, 43_200),
      priority: value.priority === undefined ? "normal" : enumString(value.priority, `channels[${index}].priority`, PRIORITIES),
      enabled: boolInt(value.enabled, true),
      toolBinding: optionalString(value.tool_binding, `channels[${index}].tool_binding`, 256),
      configJson: jsonText(value.config, `channels[${index}].config`),
      createdAt: optionalString(value.created_at, `channels[${index}].created_at`, 64) ?? now,
    };
  });
  const targetTags = requireArray(payload.target_tags ?? [], "target_tags", 5_000).map((entry, index) => {
    const value = requireObject(entry, `target_tags[${index}]`);
    return [
      requireString(value.target_id, `target_tags[${index}].target_id`, { max: 128 })!,
      requireString(value.tag_id, `target_tags[${index}].tag_id`, { max: 128 })!,
    ];
  });
  const channelTags = requireArray(payload.channel_tags ?? [], "channel_tags", 10_000).map((entry, index) => {
    const value = requireObject(entry, `channel_tags[${index}]`);
    return [
      requireString(value.channel_id, `channel_tags[${index}].channel_id`, { max: 128 })!,
      requireString(value.tag_id, `channel_tags[${index}].tag_id`, { max: 128 })!,
    ];
  });

  assertUnique(targets, "id", "targets");
  assertUnique(targets, "slug", "targets");
  assertUnique(tags, "id", "tags");
  assertUnique(tags, "slug", "tags");
  assertUnique(channels, "id", "channels");
  assertUnique(channels, "slug", "channels");

  const statements: D1PreparedStatement[] = [];
  if (mode === "replace") {
    statements.push(env.DB.prepare("UPDATE channels SET enabled = 0, updated_at = ?").bind(now));
    statements.push(env.DB.prepare("UPDATE targets SET enabled = 0, updated_at = ?").bind(now));
  }
  for (const tag of tags) {
    statements.push(env.DB.prepare(`INSERT INTO tags (id, slug, name, tag_type, created_at)
      VALUES (?, ?, ?, ?, ?)
      ON CONFLICT(id) DO UPDATE SET slug=excluded.slug, name=excluded.name, tag_type=excluded.tag_type`)
      .bind(tag.id, tag.slug, tag.name, tag.tagType, tag.createdAt));
  }
  for (const target of targets) {
    statements.push(env.DB.prepare(`INSERT INTO targets
      (id, slug, name, target_type, description, priority, enabled, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
      ON CONFLICT(id) DO UPDATE SET slug=excluded.slug, name=excluded.name,
      target_type=excluded.target_type, description=excluded.description, priority=excluded.priority,
      enabled=excluded.enabled, updated_at=excluded.updated_at`)
      .bind(target.id, target.slug, target.name, target.targetType, target.description, target.priority, target.enabled, target.createdAt, now));
  }
  for (const channel of channels) {
    statements.push(env.DB.prepare(`INSERT INTO channels
      (id, target_id, slug, name, channel_type, collector_type, url, handle, interval_minutes,
       priority, enabled, tool_binding, config_json, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
      ON CONFLICT(id) DO UPDATE SET target_id=excluded.target_id, slug=excluded.slug, name=excluded.name,
      channel_type=excluded.channel_type, collector_type=excluded.collector_type, url=excluded.url,
      handle=excluded.handle, interval_minutes=excluded.interval_minutes, priority=excluded.priority,
      enabled=excluded.enabled, tool_binding=excluded.tool_binding, config_json=excluded.config_json,
      updated_at=excluded.updated_at`)
      .bind(channel.id, channel.targetId, channel.slug, channel.name, channel.channelType,
        channel.collectorType, channel.url, channel.handle, channel.intervalMinutes, channel.priority,
        channel.enabled, channel.toolBinding, channel.configJson, channel.createdAt, now));
  }
  if (mode === "replace") {
    statements.push(env.DB.prepare("DELETE FROM target_tags"));
    statements.push(env.DB.prepare("DELETE FROM channel_tags"));
  }
  for (const [targetId, tagId] of targetTags) {
    statements.push(env.DB.prepare("INSERT OR IGNORE INTO target_tags (target_id, tag_id) VALUES (?, ?)").bind(targetId, tagId));
  }
  for (const [channelId, tagId] of channelTags) {
    statements.push(env.DB.prepare("INSERT OR IGNORE INTO channel_tags (channel_id, tag_id) VALUES (?, ?)").bind(channelId, tagId));
  }
  if (statements.length > 0) await env.DB.batch(statements);

  return {
    status: 200,
    body: {
      mode,
      synced: {
        targets: targets.length,
        channels: channels.length,
        tags: tags.length,
        target_tags: targetTags.length,
        channel_tags: channelTags.length,
      },
      synced_at: now,
    },
  };
}

function assertUnique<T extends JsonObject>(values: T[], key: keyof T, name: string): void {
  const seen = new Set<unknown>();
  for (const value of values) {
    if (seen.has(value[key])) throw new ApiError(400, "duplicate_catalog_key", `${name} contains a duplicate ${String(key)}`);
    seen.add(value[key]);
  }
}

export async function getDueChannels({ env, url }: AuthContext): Promise<ApiResponse> {
  const rawLimit = url.searchParams.get("limit");
  const limit = rawLimit === null ? 100 : requireNumber(Number(rawLimit), "limit", 1, 500);
  const now = url.searchParams.get("now") ?? new Date().toISOString();
  const targetId = url.searchParams.get("target_id");
  const rows = await env.DB.prepare(`SELECT c.*, t.slug AS target_slug, t.name AS target_name
    FROM channels c JOIN targets t ON t.id = c.target_id
    WHERE c.enabled = 1 AND t.enabled = 1
      AND (? IS NULL OR c.target_id = ?)
      AND (c.last_checked_at IS NULL OR datetime(c.last_checked_at, '+' || c.interval_minutes || ' minutes') <= datetime(?))
    ORDER BY CASE c.priority WHEN 'critical' THEN 0 WHEN 'high' THEN 1 WHEN 'normal' THEN 2 ELSE 3 END,
             COALESCE(c.last_checked_at, '1970-01-01T00:00:00Z')
    LIMIT ?`).bind(targetId, targetId, now, limit).all();
  return { status: 200, body: { channels: rows.results ?? [], as_of: now } };
}

File: intelligence/src/intelligence/models/catalog.py
"""Typed domain models for the Git-backed catalog."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Dict, List, Mapping, Optional
from uuid import UUID, uuid5


CATALOG_NAMESPACE = UUID("d7fd0429-53d1-4a72-8fd1-196af9cbe596")


class Priority(str, Enum):
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"
    CRITICAL = "critical"


def stable_id(kind: str, slug: str) -> str:
    """Return a stable identifier without storing generated IDs in YAML."""

    return str(uuid5(CATALOG_NAMESPACE, "%s:%s" % (kind, slug)))


@dataclass(frozen=True)
class Tag:
    slug: str
    name: str
    tag_type: str

    @property
    def id(self) -> str:
        return stable_id("tag", self.slug)

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "Tag":
        return cls(
            slug=str(value["slug"]),
            name=str(value["name"]),
            tag_type=str(value.get("type", "topic")),
        )

    def to_dict(self) -> Dict[str, Any]:
        return {"slug": self.slug, "name": self.name, "type": self.tag_type}

    def to_sync_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "slug": self.slug,
            "name": self.name,
            "tag_type": self.tag_type,
        }


@dataclass(frozen=True)
class Channel:
    slug: str
    name: str
    channel_type: str
    collector_type: str
    interval_minutes: int
    priority: Priority = Priority.NORMAL
    enabled: bool = True
    tier: str = "core"
    url: Optional[str] = None
    handle: Optional[str] = None
    tool_binding: Optional[str] = None
    tags: List[str] = field(default_factory=list)
    config: Dict[str, Any] = field(default_factory=dict)
    fallbacks: List[Dict[str, Any]] = field(default_factory=list)

    @property
    def id(self) -> str:
        return stable_id("channel", self.slug)

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "Channel":
        return cls(
            slug=str(value["slug"]),
            name=str(value["name"]),
            channel_type=str(value["type"]),
            collector_type=str(value["collector"]),
            interval_minutes=int(value.get("interval_minutes", 60)),
            priority=Priority(str(value.get("priority", "normal"))),
            enabled=bool(value.get("enabled", True)),
            tier=str(value.get("tier", "core")),
            url=value.get("url"),
            handle=value.get("handle"),
            tool_binding=value.get("tool_binding"),
            tags=list(value.get("tags", [])),
            config=dict(value.get("config", {})),
            fallbacks=[dict(item) for item in value.get("fallbacks", [])],
        )

    def to_dict(self) -> Dict[str, Any]:
        result: Dict[str, Any] = {
            "slug": self.slug,
            "name": self.name,
            "type": self.channel_type,
            "collector": self.collector_type,
            "interval_minutes": self.interval_minutes,
            "priority": self.priority.value,
            "enabled": self.enabled,
            "tier": self.tier,
            "tags": list(self.tags),
        }
        for key, value in (
            ("url", self.url),
            ("handle", self.handle),
            ("tool_binding", self.tool_binding),
        ):
            if value is not None:
                result[key] = value
        if self.config:
            result["config"] = dict(self.config)
        if self.fallbacks:
            result["fallbacks"] = [dict(item) for item in self.fallbacks]
        return result

    def to_sync_dict(self, target_slug: str) -> Dict[str, Any]:
        config = dict(self.config)
        config["tier"] = self.tier
        if self.fallbacks:
            config["fallbacks"] = [dict(item) for item in self.fallbacks]
        result: Dict[str, Any] = {
            "id": self.id,
            "target_id": stable_id("target", target_slug),
            "slug": self.slug,
            "name": self.name,
            "channel_type": self.channel_type,
            "collector_type": self.collector_type,
            "interval_minutes": self.interval_minutes,
            "priority": self.priority.value,
            "enabled": self.enabled,
            "config": config,
        }
        for key, value in (
            ("url", self.url),
            ("handle", self.handle),
            ("tool_binding", self.tool_binding),
        ):
            if value is not None:
                result[key] = value
        return result


@dataclass(frozen=True)
class Target:
    slug: str
    name: str
    target_type: str
    priority: Priority = Priority.NORMAL
    enabled: bool = True
    description: Optional[str] = None
    tags: List[str] = field(default_factory=list)
    channels: List[Channel] = field(default_factory=list)

    @property
    def id(self) -> str:
        return stable_id("target", self.slug)

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "Target":
        return cls(
            slug=str(value["slug"]),
            name=str(value["name"]),
            target_type=str(value["type"]),
            priority=Priority(str(value.get("priority", "normal"))),
            enabled=bool(value.get("enabled", True)),
            description=value.get("description"),
            tags=list(value.get("tags", [])),
            channels=[Channel.from_dict(item) for item in value.get("channels", [])],
        )

    def to_dict(self) -> Dict[str, Any]:
        result: Dict[str, Any] = {
            "slug": self.slug,
            "name": self.name,
            "type": self.target_type,
            "priority": self.priority.value,
            "enabled": self.enabled,
            "tags": list(self.tags),
            "channels": [channel.to_dict() for channel in self.channels],
        }
        if self.description is not None:
            result["description"] = self.description
        return result

    def to_sync_dict(self) -> Dict[str, Any]:
        result: Dict[str, Any] = {
            "id": self.id,
            "slug": self.slug,
            "name": self.name,
            "target_type": self.target_type,
            "priority": self.priority.value,
            "enabled": self.enabled,
        }
        if self.description is not None:
            result["description"] = self.description
        return result


@dataclass(frozen=True)
class Catalog:
    version: int
    tags: List[Tag]
    targets: List[Target]
    metadata: Dict[str, Any] = field(default_factory=dict)

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "Catalog":
        return cls(
            version=int(value.get("version", 1)),
            tags=[Tag.from_dict(item) for item in value.get("tags", [])],
            targets=[Target.from_dict(item) for item in value.get("targets", [])],
            metadata=dict(value.get("metadata", {})),
        )

    def to_dict(self) -> Dict[str, Any]:
        result = {
            "version": self.version,
            "tags": [tag.to_dict() for tag in self.tags],
            "targets": [target.to_dict() for target in self.targets],
        }
        if self.metadata:
            result["metadata"] = dict(self.metadata)
        return result

    def to_sync_dict(self) -> Dict[str, Any]:
        tag_ids = {tag.slug: tag.id for tag in self.tags}
        result: Dict[str, Any] = {
            "version": self.version,
            "mode": "replace",
            "tags": [tag.to_sync_dict() for tag in self.tags],
            "targets": [target.to_sync_dict() for target in self.targets],
            "channels": [
                channel.to_sync_dict(target.slug)
                for target in self.targets
                for channel in target.channels
            ],
            "target_tags": [
                {"target_id": target.id, "tag_id": tag_ids[tag_slug]}
                for target in self.targets
                for tag_slug in target.tags
            ],
            "channel_tags": [
                {"channel_id": channel.id, "tag_id": tag_ids[tag_slug]}
                for target in self.targets
                for channel in target.channels
                for tag_slug in channel.tags
            ],
        }
        if self.metadata:
            result["metadata"] = dict(self.metadata)
        return result
